zebra-network/src/meta_addr.rs

```diff
@@ -422,19 +398,90 @@ impl MetaAddr {
+    /// we should receive at least one message from a peer, or close the
+    /// connection. Therefore, if the last-seen timestamp is older than
+    /// [`constants::LIVE_PEER_DURATION`] ago, we know we should have
+    /// disconnected from it. Otherwise, we could potentially be connected to it.
+    pub fn was_recently_live(&self) -> bool {
+        if let Some(last_response) = self.last_response {
+            if let Some(elapsed) = last_response.elapsed() {
+                elapsed
+                    <= constants::LIVE_PEER_DURATION
+                        .as_secs()
+                        .try_into()
+                        .expect("unexpectedly large constant")
+            } else {
+                info!(last_response = ?self.last_response,
+                      now = ?DateTime32::now(),
+                      "unexpected future response time: assuming peer is live. Hint: has the system clock changed?"
+                );
+                // future times should be considered live
+                true
+            }
+        } else {
+            // if there is no response, it can't possibly be live
+            false
         }
     }
 
-    /// Is this address valid for outbound connections?
-    pub fn is_valid_for_outbound(&self) -> bool {
-        self.services.contains(PeerServices::NODE_NETWORK)
-            && !self.addr.ip().is_unspecified()
-            && self.addr.port() != 0
+    /// Have we recently attempted an outbound connection to this peer?
+    ///
+    /// Returns `true` if this peer was recently attempted, or has a connection
+    /// attempt in progress.
+    pub fn was_recently_attempted(&self) -> bool {
+        self.last_attempt
+            // `now` should always be later than `last_attempt`, except for tests
+            // so we can call future attempts "not recent" to simplify the code
+            .map(|last_attempt| Instant::now().checked_duration_since(last_attempt))
+            .flatten()
+            .map(|since_last_attempt| since_last_attempt <= constants::LIVE_PEER_DURATION)
+            .unwrap_or(false)
+    }
+
+    /// Have we recently had a failed connection to this peer?
+    ///
+    /// Returns `true` if this peer has recently failed.
+    pub fn was_recently_failed(&self) -> bool {
+        self.last_failure
+            // `now` should always be later than `last_failure`, except for tests
+            .map(|last_failure| Instant::now().checked_duration_since(last_failure))
+            .flatten()
+            .map(|since_last_failure| since_last_failure <= constants::LIVE_PEER_DURATION)
+            .unwrap_or(false)
+    }
+
+    /// Is this address ready for a new outbound connection attempt?
+    pub fn is_ready_for_attempt(&self) -> bool {
+        self.last_known_info_is_valid_for_outbound()
+            && !self.was_recently_live()
+            && !self.was_recently_attempted()
+            && !self.was_recently_failed()
+    }
```

Comment:
This is a core part of the security fix: skip peers that have recently been attempted, responded, or failed.